packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts

```
@@ -0,0 +1,169 @@
+      this.scopeHandler,
+      editor,
+      position,
+      direction,
+    );
+
+    for (const targetRange of targetRangesIter) {
+      if (
+        targetRangeOpposite != null &&
+        isAdjacent(targetRangeOpposite.proximal, targetRange.proximal)
+      ) {
+        yield combineScopes(targetRangeOpposite.distal, targetRange.distal);
+        targetRangeOpposite = undefined;
+      } else {
+        yield combineScopes(targetRange.proximal, targetRange.distal);
+      }
+    }
+  }
+}
+
+function combineScopes(scope1: TargetScope, scope2: TargetScope): TargetScope {
+  if (scope1.domain.isRangeEqual(scope2.domain)) {
+    return scope1;
+  }
+
+  return {
+    editor: scope1.editor,
+    domain: scope1.domain.union(scope2.domain),
+    getTargets: (isReversed) => {
+      return constructScopeRangeTarget(isReversed, scope1, scope2);
+    },
+  };
+}
+
+function* generateTargetRangesInDirection(
+  scopeHandler: ScopeHandler,
+  editor: TextEditor,
+  position: Position,
+  direction: Direction,
+): Iterable<{ proximal: TargetScope; distal: TargetScope }> {
+  let proximal, distal: TargetScope | undefined;
+
+  const generator = scopeHandler.generateScopes(editor, position, direction, {
+    allowAdjacentScopes: true,
+    skipAncestorScopes: true,
+  });
+
+  for (const scope of generator) {
+    if (proximal == null) {
+      proximal = scope;
+    }
+
+    if (distal != null) {
+      if (!isAdjacent(distal, scope)) {
+        yield { proximal, distal };
+        proximal = scope;
+      }
+    }
+
+    distal = scope;
+  }
+
+  if (proximal != null && distal != null) {
+    yield { proximal, distal };
+  }
+}
+
+function isAdjacent(scope1: TargetScope, scope2: TargetScope): boolean {
+  if (scope1.domain.isRangeEqual(scope2.domain)) {
+    return true;
+  }
+
+  const [startTarget, endTarget] = getTargetsInDocumentOrder(
+    scope1.getTargets(false)[0],
+    scope2.getTargets(false)[0],
+  );
```

do we want to throw an error if there's more than one target? Should we have a utility function `function ensureSingleTarget(scope: Scope): Target` that checks there's only one target and returns it? That worked out well for `ensureSingleEditor` I think

---

Reused our existing and shore single target. They should really never trigger with our current implementation so am fine with using an error message that might not be perfect.